compile.sh: added thread selection; added command line switches

Schwertspize opened the pull request:
closes #3014

For convinience i used shell variables.

for example:

    BUILDTYPE=Debug BRANCH=master THREADS=4 ./compile.sh

all of these can be ommitted. if THREADS is not set, it will automagically be set. if BRANCH is not set, it will be asked, but because we lack of other branches, choosen automatically. if BUILDTYPE is not set, it will be asked. if both BUILDTYPE and BRANCH are set, it will omit the greeting, because the script will then run without user input and we dont need to spam logs with a greeting right? (I assume that people who use that use it in scripts).

bearbin commented:
compile.sh: 6: compile.sh: Syntax error: newline unexpected

Please test your code.

mathwhiz1212 commented:
For me, if I run the suggested command: `BUILDTYPE=Debug BRANCH=master THREADS=4 ./compile.sh` or `BUILDTYPE=Debug BRANCH=master ./compile.sh` It will ask me for the "compile mode" and will not compile if executed from a script.

It looks way better than what I was doing :) 

```
Choose compile mode:
* (N)Normal:  Compiles normally.
              Generates the fastest build.

* (D)Debug:   Compiles in debug mode.
              Makes your console and crashes more verbose.  
              A bit slower than Normal mode. If you plan to help
              development by reporting bugs, this is preferred.
		

Note that the script will connect to the internet in order to fetch
code after this step. It will then compile your program.
```

Schwertspize commented:
Dah I noticed myself yesterday. I messed with the variables. 

You might also see that I messed with shell compatibility, I am not sure if everything is posix compatible

Schwertspize commented:
well, i fixed them, i think, and afaik this **is** POSIX compatible.

LogicParrot commented on compile.sh:
```diff
@@ -33,96 +25,190 @@ missingDepsExit ()
-# Depdendency check.
-checkPackages ()
+	# Depdendency check.
+	checkPackages ()
```
This function in nested. Is this POSIX?

Schwertspize replied:
Should be POSIX because such functions are basically just the definition of a 'comand' (like a subscript with a subshell) 

Please correct me if it's not POSIX

LogicParrot replied:
Okay. But I think it should be un-nested anyways. Nesting seems to serve no purpose here.

LogicParrot commented on compile.sh:
```diff
@@ -33,96 +25,190 @@ missingDepsExit ()
+	if [ "$BRANCH" = "" -a "$BUILDTYPE" = "" ]; then
+		# Do we already have a repo?
+		if [ -d .git ] && [ -f easyinstall.sh ] && [ -f src/BlockArea.cpp ]; then # A good enough indicator that we're in the Cuberite git repo.
+			cd ../
```
This `cd` is needed even when there is no greeting.

Schwertspize replied:
Thanks! I didn't notice that, I just thought it was a message.

Schwertspize commented:
I added $DIRECTORY with which you can specify either the path to a cuberite directory or, as git accepts it, the path to a nonexisting or empty directory. it will perform any actions to there, like cloning there. unnested the dependency check function and fixed the directory proof checks. added an additionally check before trying to git checkout (the same as we already have).

Schwertspize commented:
@LogicParrot any review? Is this mergable? 

And if this get merged, we should include documentation about this somewhere in the book and in here that says explains the usage for scripts and maybe for normal usage too (THREADS?).

LogicParrot commented:
Let me check it this weekend.

Most Gnu/Linux scripts have built-in documentation, triggered when passing a `--help` / `-h` parameter or passing the wrong parameters.

Schwertspize commented:
I know but that way we would need to parse arguments.... Anyway, should I include that if $1 is nonzero (the user wanted to Parse arguments, which could be `./compile.sh anything` or `./compile.sh -h`) it outputs a documentation?

Schwertspize commented:
@LogicParrot any news?

tigerw commented:
LogicParrot said that he'll be busy, so if it works we could merge now. We do have a very enlivening [thread](https://forum.cuberite.org/thread-2363.html) about stagnant PRs.

Schwertspize commented:
I know there is this thead, I followed it and lastly posted a note ;) 

Well, so tomorrow I will update the branch and do some basic help if you pass an argument.

LogicParrot commented:
I apologize for the delay. I have not abandoned this. I will check it as soon as I finish my exams. @sphinxc0re @tigerw please do not merge impulsively without good testing :P

I glanced this *very* quickly, and there is some code duplication, I don't like that and it is avoidable.

LogicParrot commented on compile.sh:
```diff
@@ -33,96 +25,194 @@ missingDepsExit ()
-	echo "$PROGRAMS" | while IFS='	' read EXE_NAME PACKAGE_NAME __trash__; do
-		command -v $EXE_NAME > /dev/null 2> /dev/null || printf %s " $PACKAGE_NAME"
-	done
+	osName=`uname`
```
is "Uname" POSIX? If not, this breaks non Linux compilations.

LogicParrot replied:
Perhaps 
```
osNamew=`uname || echo other`
```

LogicParrot commented on compile.sh:
```diff
@@ -33,96 +25,194 @@ missingDepsExit ()
+	elif [ "$osName" = "Linux" ]; then
+		THREADS=`nproc`;
+	else
+		wCount=`whereis nproc | awk '{print $2}' | wc -m`;
```
Overkill.
```
THREADS=`nproc 2>/dev/null || echo 2`
``

LogicParrot commented on compile.sh:
```diff
@@ -33,96 +25,194 @@ missingDepsExit ()
+	echo " --- Downloading Cuberite's source code from the $BRANCH branch..."
+
+
+	if [ -d $DIRECTORY ] && [ -d $DIRECTORY/.git ] && [ -f $DIRECTORY/easyinstall.sh ] && [ -f $DIRECTORY/src/BlockArea.cpp ]; then # A good enough indicator that we're dealing with the Cuberite git repo.
```
Please avoid this dupe.

LogicParrot commented:
This uses shell variables and not command line arguments, this is very non standard. Is there a reason for that?

Schwertspize commented:
It's way easier to parse and way easier to use :) 

But that's just my opinion 

As for most of your comments: That was original code, I did change nothing. For example, the dupe was very early in the history, I just moved it away (and copied it). Same applies to nproc or uname. (well, we don't even need uname if we handle the nproc differently. Let's see what I can do.

LogicParrot commented on compile.sh:
```diff
@@ -33,96 +25,194 @@ missingDepsExit ()
+checkRepository ()
+{
+	# Do we already have a repo?
+	if [ -d .git ] && [ -f easyinstall.sh ] && [ -f src/BlockArea.cpp ]; then # A good enough indicator that we're in the Cuberite git repo.
```
No, the dupe is new. This is the second time this line is present.

LogicParrot commented:
I submitted a PR with command line arguments. See #3255

LogicParrot commented:
>It's way easier to parse and way easier to use :) 

Parsing may be slightly easier, but I don't see how this easier **for users**.  The problem is that this is extremely uncommon and users will be confused. I prefer #3255

LogicParrot commented:
>For example, the dupe was very early in the history, I just moved it away (and copied it). Same applies to nproc or uname. (well, we don't even need uname if we handle the nproc differently. Let's see what I can do.

I do not see `nproc`, `uname`, or a douplication of the following line, in the current `compile.sh`.

```if [ -d .git ] && [ -f easyinstall.sh ] && [ -f src/BlockArea.cpp ]```

Schwertspize commented:
Do as you want to. 

These things were what was left from open issues.

LogicParrot commented:
Thanks for the effort, though :)
